pkg/resourcemanager/mysql/server/reconcile.go

```diff
@@ -87,142 +105,107 @@ func (m *MySQLServerClient) Ensure(ctx context.Context, obj runtime.Object, opts
+	if err != nil {
+		azerr := errhelp.NewAzureError(err)
 
-		skuInfo := mysql.Sku{
-			Name:     to.StringPtr(instance.Spec.Sku.Name),
-			Tier:     mysql.SkuTier(instance.Spec.Sku.Tier),
-			Capacity: to.Int32Ptr(instance.Spec.Sku.Capacity),
-			Size:     to.StringPtr(instance.Spec.Sku.Size),
-			Family:   to.StringPtr(instance.Spec.Sku.Family),
+		catchInProgress := []string{
+			errhelp.AsyncOpIncompleteError,
+			errhelp.AlreadyExists,
+		}
+		catchKnownError := []string{
+			errhelp.ResourceGroupNotFoundErrorCode,
+			errhelp.ParentNotFoundErrorCode,
+			errhelp.NotFoundErrorCode,
+			errhelp.ServiceBusy,
+			errhelp.InternalServerError,
 		}
 
-		pollURL, _, err := m.CreateServerIfValid(
-			ctx,
-			*instance,
-			labels,
-			skuInfo,
-			adminCreds.username,
-			adminCreds.password,
-			createMode,
-			hash,
-		)
-		if err != nil {
-			instance.Status.Message = errhelp.StripErrorIDs(err)
-			instance.Status.Provisioning = false
-
-			azerr := errhelp.NewAzureError(err)
-
-			catchInProgress := []string{
-				errhelp.AsyncOpIncompleteError,
-				errhelp.AlreadyExists,
-			}
-			catchKnownError := []string{
-				errhelp.ResourceGroupNotFoundErrorCode,
-				errhelp.ParentNotFoundErrorCode,
-				errhelp.NotFoundErrorCode,
-				errhelp.ServiceBusy,
-				errhelp.InternalServerError,
-			}
-
-			// handle the errors
-			if helpers.ContainsString(catchInProgress, azerr.Type) {
-				if azerr.Type == errhelp.AsyncOpIncompleteError {
-					instance.Status.PollingURL = pollURL
-				}
-				instance.Status.Message = "MySQL server exists but may not be ready"
-				instance.Status.Provisioning = true
-				return false, nil
-			}
-
-			if helpers.ContainsString(catchKnownError, azerr.Type) {
-				return false, nil
+		// handle the errors
+		if helpers.ContainsString(catchInProgress, azerr.Type) {
+			if azerr.Type == errhelp.AsyncOpIncompleteError {
+				instance.Status.PollingURL = pollURL
 			}
+			instance.Status.SetProvisioning("MySQL server exists but may not be ready")
+			return false, nil
+		}
 
-			// serious error occurred, end reconciliation and mark it as failed
-			instance.Status.Message = errhelp.StripErrorIDs(err)
-			instance.Status.Provisioned = false
-			instance.Status.FailedProvisioning = true
-			return true, nil
+		if helpers.ContainsString(catchKnownError, azerr.Type) {
+			instance.Status.SetProvisioning(errhelp.StripErrorIDs(err))
+			return false, nil
 		}
 
-		instance.Status.PollingURL = pollURL
-		instance.Status.Message = "request submitted to Azure"
+		// serious error occurred, end reconciliation and mark it as failed
+		instance.Status.SetFailedProvisioning(errhelp.StripErrorIDs(err))
+		return true, nil
 	}
+
+	instance.Status.SetPollingURL(pollURL, api.PollingURLKindCreateOrUpdate)
+	instance.Status.SetProvisioning("request submitted to Azure")
+
 	return false, nil
 }
 
+func (m *MySQLServerClient) calculateHash(spec azurev1alpha2.MySQLServerSpec, secret map[string][]byte) string {
+	return helpers.Hash256(struct {
```

Comment:
Cunning!

Reply:
I thought it was a bit hacky but didn't have a better way, and it does do the job.